### PR TITLE
build: add lint and fmt steps for shell scripts

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

- add a Docker-based pre-commit hook
- check through CI

[DET-7566]

## Commentary (optional)

indent w/ 2 space or 4? (🚲) 4 results in less changes (500loc vs 300)
```
     Generic flags
         -version
             Show version and exit.

         -l, --list
             List files whose formatting differs from shfmt's.

         -w, --write
             Write result to file instead of stdout.

         -d, --diff
             Error with a diff when the formatting differs.

             The diff uses color when the output is a terminal. To never use color, set a non-empty NO_COLOR or TERM=dumb. To always use color, set a non-
             empty FORCE_COLOR.

         -s, --simplify
             Simplify the code.

         -mn, --minify
             Minify the code to reduce its size (implies -s).

     Parser flags
         -ln, --language-dialect <str>
             Language dialect (bash/posix/mksh/bats, default auto).

             When set to auto, the language is detected from the input filename, as long as it has a shell extension like foo.mksh. Otherwise, if the input
             begins with a shell shebang like #!/bin/sh, that's used instead. If neither come up with a result, bash is used as a fallback.

             The filename extension .sh is a special case: it implies posix, but may be overriden by a valid shell shebang.

         -p, --posix
             Shorthand for -ln=posix.

         -filename str
             Provide a name for the standard input file.

     Printer flags
         -i, --indent <uint>
             Indent: 0 for tabs (default), >0 for number of spaces.

         -bn, --binary-next-line
             Binary ops like && and | may start a line.

         -ci, --switch-case-indent
             Switch cases will be indented.

         -sr, --space-redirects
             Redirect operators will be followed by a space.

         -kp, --keep-padding
             Keep column alignment paddings.

         -fn, --func-next-line
             Function opening braces are placed on a separate line.
```

## Checklist
- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-7566]: https://determinedai.atlassian.net/browse/DET-7566?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: webui/react/src/services/api-ts-sdk/git_push.sh [hunk 1]
```diff
@@ -28,15 +28,15 @@ git init
...
 git add .
 
-# Commits the tracked changes and prepares them to be pushed to a remote repository.
+# Commits the tracked changes and prepares them to be pushed to a remote repository. 
 git commit -m "$release_note"
 
 # Sets the new remote
-git_remote=$(git remote)
+git_remote=`git remote`
 if [ "$git_remote" = "" ]; then # git remote not defined
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply webui/react/src/services/api-ts-sdk/git_push.sh:36]: you might be looking at a commit that's reverting changes. (avoiding changes to generated shell scripts due to collision with another job we run)

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -24,6 +24,7 @@ repos:
         description: Format shell scripts with shfmt
         entry: mvdan/shfmt:v3
         language: docker_image
+        exclude: ^(webui/react/src/services/api-ts-sdk/|other/folder/)
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply .pre-commit-config.yaml:27]: TODO: remove other/folder

[the remark this answers is not in the document]
Comment (hamidzr) [reply .pre-commit-config.yaml:27]: also, that looks great I didn't know about it. let's add that (separately)

File: webui/react/src/services/api-ts-sdk/git_push.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (dannysauer) [reply webui/react/src/services/api-ts-sdk/git_push.sh:36]: Ah, duh. That's why I didn't notice it on the full PR review.

**Comment by hamidzr** [comment]
thanks for the review. I'll wait a bit to see if @stoksc has any comments

File: tools/k8s/fetch-creds.sh [hunk 3]
```diff
@@ -45,9 +45,9 @@ print_help () {
     echo ""
     echo "  - Run an unmodified k8s application via docker run:"
     echo ""
-    echo "      docker run \\"
-    echo "          -v OUTDIR:/var/run/secrets/kubernetes.io/serviceaccount \\"
-    echo "          --env-file OUTDIR/docker-env-file \\"
+    echo '      docker run \'
+    echo '          -v OUTDIR:/var/run/secrets/kubernetes.io/serviceaccount \'
+    echo '          --env-file OUTDIR/docker-env-file \'
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply tools/k8s/fetch-creds.sh:50]: hopefully, it doesn't break anything either. I'm assuming changes that shfmt makes will not result in any change of behavior whatsoever 🤞🏼

**Review by stoksc — APPROVED** [review]
shell scripts look great to me, though take @dannysauer 's word because i dont know github ci at all

File: tools/k8s/fetch-creds.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (dannysauer) [reply tools/k8s/fetch-creds.sh:50]: It only did a couple things I wouldn't do, like unquoting vars on the left hand side of comparisons and removing line continuation chars on multi-line command replacements. But as far as I know, everything it did is technically valid with "modern" shells (bash/zsh).

We're really not directly testing our shell scripts. I guess that's something we could think about improving in the future.

**Comment by dzhu** [comment]
Any chance we can add in `-ci -bn` (indent switch cases and move binary operators at the beginning of lines)?

(I would also like to note that I'm personally used to 2-space indents, but since that already got actively decided against, I'll leave it at that.)

**Comment by hamidzr** [comment]
I think I prefer 2 spaces as well.. I don't know. do you wanna start a poll in #engineering : )?

File: .pre-commit-config.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (hamidzr) [reply .pre-commit-config.yaml]: https://determinedai.atlassian.net/browse/DET-8645

**Comment by hamidzr** [comment]
I'll set up the auto-merge to establish a baseline. it'll be easy to flip the flag once we have it automated with this pr

**Pull request merged**